File: LeetCodeSolutions/LeetCode_0067.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = ''
        la = len(a)
        lb = len(b)
        l = max(la, lb)
        a = a.rjust(l, '0')
        b = b.rjust(l, '0')
        carry = 0
        i = l - 1
        while i > -1:
            tmp, carry = self.stradd(a[i], b[i], carry)
            res = tmp + res
            i -= 1
        if carry:
            res = '1' + res
        return res

    def stradd(self, a, b, carry):
        if (a == '0' and b == '0'):
            if carry:
                return '1', 0
            else:
                return '0', 0
        if (a == '0' and b == '1') or (a == '1' and b == '0'):
            if carry:
                return '0', 1
            else:
                return '1', 0
        if (a == '1' and b == '1'):
            if carry:
                return '1', 1
            else:
                return '0', 1

File: LeetCodeSolutions/LeetCode_0067.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int i = a.size() - 1, j = b.size() - 1;
        string ret;
        int carry = 0;
        while(i >= 0 || j >= 0 || carry > 0){
            char a_ = i >= 0 ? a[i]: '0';
            char b_ = j >= 0 ? b[j]: '0';
            if(i >= 0)--i;
            if(j >= 0)--j;
            if (a_ != b_){
                if(carry) {
                    ret += '0';
                    carry = 1;
                }
                else{ret += '1'; carry = 0;}
            }
            else{
                if(a_ == '0'){
                    ret += carry? '1' : '0';
                    carry = 0;
                }
                else{
                    ret += carry? '1': '0';
                    carry = 1;
                }
            }
        }
        return {ret.rbegin(), ret.rend()};

    }
};